From cb5e252cc94e79f24c9ce26047b39ca7f173d978 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?Marcin=20Wcis=C5=82o?= <marcin.wcislo@conclusive.pl>
Date: Thu, 16 Jul 2026 20:04:54 +0200
Subject: [PATCH] net: dst: introduce dst->dev_rcu
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

jira VULN-162719
cve-pre CVE-2025-40170
commit-author Eric Dumazet <edumazet@google.com>
commit caedcc5b6df1b2e2b5f39079e3369c1d4d5c5f50
upstream-diff Used CentOS 9 backport
  ca6810ffc83408a1dd36676447c35d0348ef0571 as a clean cherry-pick base.
  Used RH_KABI_REPLACE in `struct dst_entry' to silence the kABI
  checker. Originally the `struct net_device *dev' field was changed
  into a union of `struct net_device *dev' and `struct net_device __rcu
  *dev_rcu'. The latter has the same size and alignment as the former
  (this is actually checked statically by the RH_KABI_REPLACE macro),
  with the `__rcu' annotation serving only at compile time and being
  invisible to the binary output. The `dev_rcu' is therefore just an
  alias for the `dev' field.

Followup of commit 88fe14253e18 ("net: dst: add four helpers
to annotate data-races around dst->dev").

We want to gradually add explicit RCU protection to dst->dev,
including lockdep support.

Add an union to alias dst->dev_rcu and dst->dev.

Add dst_dev_net_rcu() helper.

Fixes: 4a6ce2b6f2ec ("net: introduce a new function dst_dev_put()")
	Signed-off-by: Eric Dumazet <edumazet@google.com>
	Reviewed-by: David Ahern <dsahern@kernel.org>
Link: https://patch.msgid.link/20250828195823.3958522-2-edumazet@google.com
	Signed-off-by: Jakub Kicinski <kuba@kernel.org>
(cherry picked from commit d566795fc6a28cf9457511b428c86bb783ac7d7c)
	Signed-off-by: Marcin Wcisło <marcin.wcislo@conclusive.pl>
---
 include/net/dst.h | 18 +++++++++++++-----
 net/core/dst.c    |  2 +-
 net/ipv4/route.c  |  2 +-
 3 files changed, 15 insertions(+), 7 deletions(-)

diff --git a/include/net/dst.h b/include/net/dst.h
index d1fcd3e5ec5f..73d797809899 100644
--- a/include/net/dst.h
+++ b/include/net/dst.h
@@ -26,7 +26,12 @@
 struct sk_buff;
 
 struct dst_entry {
-	struct net_device       *dev;
+	RH_KABI_REPLACE(
+		struct net_device       *dev,
+		union {
+			struct net_device       *dev;
+			struct net_device __rcu *dev_rcu;
+		})
 	struct  dst_ops	        *ops;
 	unsigned long		_metrics;
 	unsigned long           expires;
@@ -586,9 +591,12 @@ static inline struct net_device *dst_dev(const struct dst_entry *dst)
 
 static inline struct net_device *dst_dev_rcu(const struct dst_entry *dst)
 {
-	/* In the future, use rcu_dereference(dst->dev) */
-	WARN_ON_ONCE(!rcu_read_lock_held());
-	return READ_ONCE(dst->dev);
+	return rcu_dereference(dst->dev_rcu);
+}
+
+static inline struct net *dst_dev_net_rcu(const struct dst_entry *dst)
+{
+	return dev_net_rcu(dst_dev_rcu(dst));
 }
 
 static inline struct net_device *skb_dst_dev(const struct sk_buff *skb)
@@ -608,7 +616,7 @@ static inline struct net *skb_dst_dev_net(const struct sk_buff *skb)
 
 static inline struct net *skb_dst_dev_net_rcu(const struct sk_buff *skb)
 {
-	return dev_net_rcu(skb_dst_dev(skb));
+	return dev_net_rcu(skb_dst_dev_rcu(skb));
 }
 
 struct dst_entry *dst_blackhole_check(struct dst_entry *dst, u32 cookie);
diff --git a/net/core/dst.c b/net/core/dst.c
index 3bef7348fec1..cc01dfb55707 100644
--- a/net/core/dst.c
+++ b/net/core/dst.c
@@ -152,7 +152,7 @@ void dst_dev_put(struct dst_entry *dst)
 		dst->ops->ifdown(dst, dev, true);
 	dst->input = dst_discard;
 	dst->output = dst_discard_out;
-	WRITE_ONCE(dst->dev, blackhole_netdev);
+	rcu_assign_pointer(dst->dev_rcu, blackhole_netdev);
 	netdev_ref_replace(dev, blackhole_netdev, &dst->dev_tracker,
 			   GFP_ATOMIC);
 }
diff --git a/net/ipv4/route.c b/net/ipv4/route.c
index 2e29bf3f1c01..3cf3f5b2aaa8 100644
--- a/net/ipv4/route.c
+++ b/net/ipv4/route.c
@@ -1035,7 +1035,7 @@ static void __ip_rt_update_pmtu(struct rtable *rt, struct flowi4 *fl4, u32 mtu)
 		return;
 
 	rcu_read_lock();
-	net = dev_net_rcu(dst->dev);
+	net = dst_dev_net_rcu(dst);
 	if (mtu < ip_rt_min_pmtu) {
 		lock = true;
 		mtu = min(old_mtu, ip_rt_min_pmtu);
-- 
2.51.0

